### PR TITLE
Spark: Inject `DataSourceV2Relation` when missing

### LINKED ISSUE #7226: Structured streaming writes to partitioned table fails when spark.sql.extensions is set to IcebergSparkSessionExtensions
### Apache Iceberg version

1.2.0 (latest release)

### Query engine

Spark

### Please describe the bug 🐞

According to the documentation, when using Iceberg, one should set `spark.sql.extensions` to `org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions`, but setting this property seems to cause an exception to be thrown when trying to write to an Iceberg table using Spark structured streaming.

The exception that is thrown is:

```
Exception in thread "main" org.apache.spark.sql.streaming.StreamingQueryException: days(ts) is not currently supported
=== Streaming Query ===
Identifier: [id = cfb83943-cd87-4c84-bf25-a290e8891e19, runId = ddf71690-7e5d-41f6-8a8e-84c425683a26]
Current Committed Offsets: {}
Current Available Offsets: {MemoryStream[ts#3,a#4,b#5]: 0}

Current State: ACTIVE
Thread State: RUNNABLE

Logical Plan:
WriteToMicroBatchDataSource spark_catalog.default.test_iceberg_table, cfb83943-cd87-4c84-bf25-a290e8891e19, [path=test_iceberg_table, fanout-enabled=true, checkpointLocation=/tmp/spark-checkpoint-16659193840247202419], Append
+- StreamingDataSourceV2Relation [ts#3, a#4, b#5], org.apache.spark.sql.execution.streaming.MemoryStreamScanBuilder@4533d1cf, MemoryStream[ts#3,a#4,b#5]
```

Code to reproduce:

```scala
package com.example

import org.apache.spark.sql.execution.streaming.MemoryStream
import org.apache.spark.sql.streaming.Trigger
import org.apache.spark.sql.{Encoders, SparkSession}

import java.nio.file.Files
import java.sql.Timestamp

case class Bla(ts: Timestamp, a: String, b: Double)

object MinEx {
  def main(args: Array[String]): Unit = {
    val warehouseDir = Files.createTempDirectory("spark-warehouse-iceberg-").toString
    val checkpointDir = Files.createTempDirectory("spark-checkpoint-").toString
    val spark = SparkSession.builder()
      .config("spark.sql.catalog.spark_catalog", "org.apache.iceberg.spark.SparkSessionCatalog")
      .config("spark.sql.catalog.spark_catalog.type", "hadoop")
      .config("spark.sql.catalog.spark_catalog.warehouse", warehouseDir)
      .config("spark.sql.warehouse.dir", warehouseDir)
      .config("spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions")
      .appName("BugRepro")
      .master("local[*]")
      .enableHiveSupport()
      .getOrCreate()

    spark.sql("create table test_iceberg_table(ts timestamp, a string, b double) using iceberg partitioned by (days(ts))")

    implicit val sqlContext = spark.sqlContext
    implicit val encoder = Encoders.product[Bla]
    val memStream = MemoryStream[Bla]
    val now = System.currentTimeMillis()
    val day = 86400000
    memStream.addData(List(
      Bla(new Timestamp(now), "test", 12.34),
      Bla(new Timestamp(now - 1 * day), "test 1d", 33.34),
      Bla(new Timestamp(now - 3 * day), "test 3d", 44.34),
      Bla(new Timestamp(now - 2 * day), "test 2d", 55.34),
    ))

    memStream.toDF()
      .writeStream
      .format("iceberg")
      .outputMode("append")
      .option("path", "test_iceberg_table")
      .option("fanout-enabled", true)
      .option("checkpointLocation", checkpointDir)
      .trigger(Trigger.Once())
      .start()
      .awaitTermination()
  }
}
```

The code works as expected when the statement that configures `spark.sql.extensions` is commented out.

### PR DESCRIPTION
**Pull request by Fokko**
When you start a structured streaming query using `.start()`, there will be no `DataSourceV2Relation` reference. Therefore the catalog functions won't be available, resulting in:

```
Exception in thread "main" org.apache.spark.sql.streaming.StreamingQueryException: days(ts) is not currently supported
```

When this is missing, we'll just create one since we know the table.

Resolves #7226

### CODE REVIEW

**Comment by Marcus-Rosti** [comment]
Is this fix only for 3.4+?

**Comment by Fokko** [comment]
@Marcus-Rosti I first wanted to get some feedback on this before also backporting this to older versions of Spark

**Comment by aokolnychyi** [comment]
I should have some time to review this week. Sorry for the delay!

**Comment by aokolnychyi** [comment]
It seems like we are trying to fix a bug in Spark, which is beyond Iceberg control. While I don't mind that cause we would have to wait for a new Spark release otherwise, it would be nice to look for a proper fix in Spark that would work without Iceberg extensions.

Let me take a closer look at the Spark side.

File: spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SetMissingRelation.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * the relation is empty, and the catalog functions will be missing.
+ * The catalog functions are required when a table is partitioned with a transform.
+ */
+case class SetMissingRelation(spark: SparkSession) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan match {
+    case p: WriteToMicroBatchDataSource if p.relation.isEmpty =>
+      import spark.sessionState.analyzer.CatalogAndIdentifier
+      val originalMultipartIdentifier = spark.sessionState.sqlParser
+        .parseMultipartIdentifier(p.table.name())
```
Comment (aokolnychyi) [inline spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SetMissingRelation.scala:39]: I have doubts that this is generally safe as we assume `table.name()` would include the catalog name. That's currently the case but it feels fragile. I mentioned a potential fix on the Spark side [here](https://github.com/apache/iceberg/issues/7226#issuecomment-1687307732). Let me come back with fresh eyes tomorrow.

@Fokko, let me know what you think on fixing it in Spark.

Comment (Fokko) [reply spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SetMissingRelation.scala:39]: I see your concern, but as you can see in the test it works quite well. In the smoke test it passes for `table`, `catalog.schema.table`, and with `s3://bucket/wh/path`. Fixing it in Spark could also work, but then I need more pointers on where to start. I looked into the comment on the issue, but it wasn't directly obvious to me. I think we should fix this, looking at how many people are bumping into this.

Comment (aokolnychyi) [reply spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SetMissingRelation.scala:39]: I'd say let's just use the new API in Spark and don't worry about it. I think you already updated the docs to cover that.

**Pull request closed without merging**